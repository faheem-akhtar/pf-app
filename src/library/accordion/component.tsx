import { FunctionComponent } from 'react';

export const AccordionComponent: FunctionComponent = (props): JSX.Element => {
  return <>{props.children}</>;
};
