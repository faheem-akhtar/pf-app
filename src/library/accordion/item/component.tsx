import { FunctionComponent, useState } from 'react';

import { IconThinChevronDownTemplate } from 'components/icon';
import { domClassMerge } from 'helpers/dom/class-merge';

import styles from './accordion-item.module.scss';
import { AccordionItemComponentPropsInterface } from './component-props.interface';

export const AccordionItemComponent: FunctionComponent<AccordionItemComponentPropsInterface> = ({
  expanded = false,
  children,
  title,
}) => {
  const [isExpanded, setIsExpanded] = useState(false || expanded);

  return (
    <div className={styles.container}>
      <header className={styles.header} onClick={(): void => setIsExpanded((isExpanded) => !isExpanded)}>
        <p className={styles.title}>{title}</p>
        <IconThinChevronDownTemplate
          class={domClassMerge(styles.icon, {
            [styles['icon--expanded']]: isExpanded,
          })}
          clipped
        />
      </header>

      {isExpanded && <div className={styles.panel}>{children}</div>}
    </div>
  );
};
