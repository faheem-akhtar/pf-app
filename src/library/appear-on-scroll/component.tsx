import { useEffect, useRef, useState } from 'react';

import { ElementService } from 'services/element/service';

import { AppearOnScrollComponentPropsInterface } from './component-props.interface';
import { appearOnScrollObserve } from './observe';
import { appearOnScrollOnStatusChange } from './on-status-change';
import { appearOnScrollOnVisibilityChange } from './on-visibility-change';
import { AppearOnScrollStatusEnum } from './status.enum';
import { AppearOnScrollTemplate } from './template';

export const AppearOnScrollComponent = (props: AppearOnScrollComponentPropsInterface): JSX.Element => {
  const wrapper = useRef<HTMLDivElement>(null);
  const prevStatusRef = useRef<AppearOnScrollStatusEnum | null>(null);
  const [status, setStatus] = useState<AppearOnScrollStatusEnum | null>(null);

  // On mount
  useEffect(() => {
    const element = wrapper.current as HTMLDivElement;

    const wrapperClientRec = ElementService.getBoundingClientRect(element);
    const wrapperInView = wrapperClientRec.y >= 0 && wrapperClientRec.y < document.body.clientHeight;
    // set the initial status on mount
    setStatus(wrapperInView ? AppearOnScrollStatusEnum.WRAPPER_IN_VIEW : AppearOnScrollStatusEnum.WRAPPER_OUTSIDE_VIEW);

    // observe the target element, when it's visibility change - call onVisibilityChange
    return appearOnScrollObserve(element, (visible) => appearOnScrollOnVisibilityChange(visible, setStatus));
  }, []);

  if (prevStatusRef.current !== status) {
    appearOnScrollOnStatusChange(props, status);
  }
  prevStatusRef.current = status;

  return (
    <div ref={wrapper}>
      <AppearOnScrollTemplate status={status} className={props.className}>
        {props.children}
      </AppearOnScrollTemplate>
    </div>
  );
};
