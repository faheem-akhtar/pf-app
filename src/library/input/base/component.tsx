import { FunctionComponent, useState } from 'react';

import { InputBaseComponentPropsInterface } from './component-props.interface';
import { InputBaseTemplate } from './template';
import { InputBaseTemplatePropsBaseInterface } from './template-props-base.interface';

export const InputBaseComponent: FunctionComponent<InputBaseComponentPropsInterface> = ({
  onChange,
  onFocus,
  onBlur,
  value,
  ...rest
}) => {
  const [isFocused, setIsFocused] = useState(false);

  /*
   * Handling input onChange event
   */
  const handleChange: InputBaseTemplatePropsBaseInterface['onChange'] = (e) => {
    const newValue = (e.target as HTMLInputElement).value;
    onChange(newValue);
  };

  /*
   * Handling input onFocus event
   */
  const handleFocus: InputBaseTemplatePropsBaseInterface['onFocus'] = (e) => {
    setIsFocused(true);

    if (onFocus) {
      onFocus(e as React.FocusEvent<HTMLInputElement>);
    }
  };

  /*
   * Handling input onBlur event
   */
  const handleBlur: InputBaseTemplatePropsBaseInterface['onBlur'] = (e) => {
    setIsFocused(false);

    if (onBlur) {
      onBlur(e as React.FocusEvent<HTMLInputElement>);
    }
  };

  return (
    <InputBaseTemplate
      value={value}
      focus={isFocused}
      onChange={handleChange}
      onFocus={handleFocus}
      onBlur={handleBlur}
      {...rest}
    />
  );
};
