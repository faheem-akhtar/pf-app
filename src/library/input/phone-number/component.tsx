import React, { FunctionComponent, useRef, useState } from 'react';

import { CountrySortEnum } from 'enums/country/sort.enum';
import { InputBaseTemplatePropsBaseInterface } from '../base/template-props-base.interface';
import { InputPhoneNumberComponentPropsInterface } from './component-props.interface';
import { InputPhoneNumberTemplate } from './template';
import { SkeletonTemplate } from 'library/skeleton/template';

import { useApiCountryPhoneCodes } from 'api/country-phone-codes/hook';
import { useWindowMouseDown } from 'helpers/window/mouse-down.hook';

export const InputPhoneNumberComponent: FunctionComponent<InputPhoneNumberComponentPropsInterface> = ({
  onBlur,
  onFocus,
  ...rest
}): JSX.Element => {
  const [isVisible, setIsVisible] = useState(false);
  const [isFocused, setIsFocused] = useState(false);
  const rootRef = useRef(null);
  const countriesListResponse = useApiCountryPhoneCodes(CountrySortEnum.priority);
  const countryList = countriesListResponse.ok ? countriesListResponse.data : [];

  useWindowMouseDown({
    shouldListen: isVisible,
    ignoreElementRef: rootRef,
    onWindowMouseDown: () => {
      setIsVisible(false);
    },
  });

  const onClickToggle = (): void => setIsVisible(!isVisible);

  /*
   * Handling input onFocus event
   */
  const handleFocus: InputBaseTemplatePropsBaseInterface['onFocus'] = (e) => {
    setIsFocused(true);

    if (onFocus) {
      onFocus(e as React.FocusEvent<HTMLInputElement>);
    }
  };

  /*
   * Handling input onBlur event
   */
  const handleBlur: InputBaseTemplatePropsBaseInterface['onBlur'] = (e) => {
    setIsFocused(false);

    if (onBlur) {
      onBlur(e as React.FocusEvent<HTMLInputElement>);
    }
  };

  if (!countryList.length) return <SkeletonTemplate />;

  return (
    <InputPhoneNumberTemplate
      onClickToggle={onClickToggle}
      visible={isVisible}
      rootRef={rootRef}
      onFocus={handleFocus}
      onBlur={handleBlur}
      focus={isFocused}
      countryList={countryList}
      {...rest}
    />
  );
};
