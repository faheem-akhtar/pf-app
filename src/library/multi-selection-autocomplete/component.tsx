import { useEffect, useRef, useState } from 'react';

import { multiSelectionAutocompleteMakeKeyboardAccessibilityHandler } from './make-keyboard-accessibility-handler';
import { useWindowMouseDownHook } from 'helpers/hook/window-mouse-down.hook';

import { KeyboardKeyEnum } from 'enums/keyboard/key.enum';
import { MultiSelectionAutocompleteComponentPropsInterface } from './component-props.interface';
import { MultiSelectionAutocompleteExtensionInterface } from './extension.interface';
import { MultiSelectionAutocompleteTemplate } from './template';

// TODO-FE[TPNX-1968] Test MultiSelectionAutocompleteComponent

export const MultiSelectionAutocompleteComponent = <T extends unknown>(
  props: MultiSelectionAutocompleteComponentPropsInterface<T>
): JSX.Element => {
  /* Refs */
  const rootRef = useRef(null);
  // It is used to keep focus inside the component while navigating up and down with arrows
  const hiddenInputRef = useRef<HTMLInputElement>(null);
  const inputRef = useRef<HTMLInputElement>(null);

  /* State */
  const [isActive, setIsActive] = useState(false);
  const [suggestions, setSuggestions] = useState<T[]>([]);
  const [inputValue, setInputValueInternal] = useState('');
  const [isLoading, setIsLoading] = useState(false);
  const [selectedItemIndex, setSelectedItemIndex] = useState<number | null>(null);

  const setInputValue = (value: string): void => {
    setInputValueInternal(value);
    if (props.onInputValueChange) {
      props.onInputValueChange(value);
    }
  };

  /* Helpers */
  const addInputFocus = (): void => {
    inputRef.current?.select();
  };

  /**
   * Removes the focus from input and keeps the hidden input focused,
   * so the keydown event can be received by the root component
   */
  const removeInputFocus = (): void => {
    hiddenInputRef.current?.focus();
    inputRef.current?.blur();
  };

  /**
   * Deactivate the component
   */
  const deactivate = (): void => {
    setIsActive(false);
    setInputValue('');
    setSelectedItemIndex(null);
    removeInputFocus();
    setSuggestions([]);
  };

  /**
   * Handle suggestions promise
   * @param promise will eventually return suggestions
   */
  const handleSuggestionsResult = (promise: Promise<T[]>): void => {
    setIsLoading(true);

    promise
      .then(setSuggestions)
      .catch((e) => {
        // eslint-disable-next-line no-console
        console.error(`MultiSelectionAutocompleteComponent.handleSuggestionsResult promise error`, e);
      })
      .then(() => setIsLoading(false));
  };

  /**
   * Add item and deactivate
   * @param item
   */
  const addItemAndDeactivate = (item: T): void => {
    deactivate();
    props.onAddItem(item);
  };

  /**
   * Handle input value change
   */
  const handleInputChange = (value: string): void => {
    setInputValue(value);
    handleSuggestionsResult(props.onInputChange(value));
  };

  /**
   * On input keydown
   * @param e KeyboardEvent
   */
  const onInputKeyDown = (e: KeyboardEvent): void => {
    // deactivate widget if user pressed tab during input is in focus
    if (e.key === KeyboardKeyEnum.tab) {
      deactivate();
    }
  };

  /**
   * On suggestion click
   * @param item
   */
  const suggestionOnClick = (item: T): void => {
    deactivate();
    props.onAddItem(item);
  };

  const extensionInterfaceProps: MultiSelectionAutocompleteExtensionInterface<T> = {
    addItem: props.onAddItem,
    deactivate,
    clearInput: () => handleInputChange(''),
    inputValue,
  };

  /* Effects */
  // deactivate on windows mouse down outside of component
  useWindowMouseDownHook({
    shouldListen: isActive,
    ignoreElementRef: rootRef,
    onWindowMouseDown: (e) => {
      if (props.onWindowMouseDownOutsideAutocomplete) {
        props.onWindowMouseDownOutsideAutocomplete({ e, ...extensionInterfaceProps });
      }
      deactivate();
    },
  });

  // focus input when component becomes active
  useEffect(() => {
    isActive && addInputFocus();
  }, [isActive]);

  return (
    <MultiSelectionAutocompleteTemplate
      chipsOnTheBottom={props.chipsOnTheBottom}
      className={props.className}
      chipClassName={props.chipClassName}
      value={props.value}
      placeholder={props.placeholder}
      onItemCrossClick={(item, isMore): void => (isMore ? props.onItemGroupRemoveClick() : props.onItemRemove(item))}
      getTitle={props.getTitle}
      getChipTitle={props.getChipTitle}
      renderNoSuggestions={props.renderNoSuggestions}
      suggestionIcon={props.suggestionIcon}
      isLoading={isLoading}
      isActive={isActive}
      inputRef={inputRef}
      inputValue={inputValue}
      onInputChange={(): void => handleInputChange(inputRef?.current?.value || '')}
      onInputKeyDown={onInputKeyDown}
      onInputCrossButtonClick={deactivate}
      suggestions={suggestions}
      selectedItemIndex={selectedItemIndex}
      suggestionOnClick={suggestionOnClick}
      rootRef={rootRef}
      hiddenInputRef={hiddenInputRef}
      onInputFocus={(): void => {
        setSelectedItemIndex(null);
        handleSuggestionsResult(props.onFocus(inputValue));
      }}
      onInputKeyPress={(e: KeyboardEvent): void =>
        props.onInputKeyPress && props.onInputKeyPress({ e, ...extensionInterfaceProps })
      }
      onRootClick={(): void => {
        if (!isActive) {
          setIsActive(true);
        }
      }}
      onRootKeyDown={multiSelectionAutocompleteMakeKeyboardAccessibilityHandler({
        selectedItemIndex,
        suggestions,
        removeInputFocus,
        addInputFocus,
        deactivate,
        setSelectedItemIndex,
        addItemAndDeactivate,
      })}
    />
  );
};
