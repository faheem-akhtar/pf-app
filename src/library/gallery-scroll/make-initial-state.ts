import { GalleryScrollBrowseEnum } from './browse.enum';
import { galleryScrollComputeItemPositionsXUpdate } from './compute-item-positions-x-update';
import { GalleryScrollItemInterface } from './item.interface';
import { GalleryScrollStateInterface } from './state.interface';

export const galleryScrollMakeInitialState = (
  items: GalleryScrollItemInterface[],
  isRtl: boolean
): GalleryScrollStateInterface => ({
  isDragging: false,
  activeIndex: 0,
  itemsPositionsX: items.reduce((acc, item, index) => {
    if (acc[index] !== undefined) {
      return acc;
    }
    return { ...acc, [index]: 100 };
  }, galleryScrollComputeItemPositionsXUpdate(items.length, isRtl, 0)),
  pointerPositionStart: null,
  pointerPositionCurrent: null,
  isTouched: false,
  browse: GalleryScrollBrowseEnum.none,
});
