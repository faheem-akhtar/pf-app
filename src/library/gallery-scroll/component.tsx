import { useReducer, useRef, useState } from 'react';

import { functionNoop } from 'helpers/function/noop';
import { galleryScrollGetSlidingStyles } from './get-sliding-styles';
import { galleryScrollMakeInitialState } from './make-initial-state';
import { galleryScrollMakeReducer } from './make-reducer';
import { useGalleryScrollEffects } from './effects.hook';

import { GalleryScrollComponentPropsInterface } from './component-props.interface';
import { GalleryScrollTemplate } from './template';

export const GalleryScrollComponent = (props: GalleryScrollComponentPropsInterface): JSX.Element => {
  const galleryRef = useRef<HTMLDivElement>(null);
  const numberOfImages = props.items.length;
  const [reducer] = useState(() => galleryScrollMakeReducer(props.isRtl, numberOfImages));
  const [initialReducerState] = useState(() => galleryScrollMakeInitialState(props.items, props.isRtl));

  const [state, dispatch] = useReducer(reducer, initialReducerState);
  useGalleryScrollEffects(state.pointerPositionStart !== null, dispatch);

  const onStart = (positionX: number): void => {
    if (!galleryRef.current) {
      return;
    }
    const galleryRect = galleryRef.current.getBoundingClientRect();

    dispatch({
      type: 'start',
      positionX,
      galleryLeft: galleryRect.left,
      galleryRight: galleryRect.right,
    });
  };

  const hasMultipleImages = numberOfImages > 1;

  const itemsToRender = props.items.map((item, i) => ({
    ...item,
    style: hasMultipleImages ? galleryScrollGetSlidingStyles(state, i, numberOfImages) : null,
  }));

  return (
    <GalleryScrollTemplate
      containerRef={galleryRef}
      className={props.className}
      onMouseDown={
        hasMultipleImages
          ? (e): void => {
              e.preventDefault();
              onStart(e.clientX);
              props.onTouch();
            }
          : functionNoop
      }
      onTouchStart={
        hasMultipleImages
          ? (e): void => {
              onStart(e.changedTouches[0].pageX);
              props.onTouch();
            }
          : functionNoop
      }
      items={itemsToRender}
      isTouched={state.isTouched}
      activeIndex={state.activeIndex}
    />
  );
};
