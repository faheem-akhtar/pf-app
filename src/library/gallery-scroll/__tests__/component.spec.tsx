/**
 * @jest-environment jsdom
 */

import { fireEvent, render } from '@testing-library/react';

import { galleryScrollMakeInitialState } from '../make-initial-state';
import { reactMockUseReducer } from 'mocks/react/mock-use-reducer';
import { windowMockRemoveEventListener } from 'mocks/window/mock-remove-event-listener';

import { GalleryScrollComponent } from '../component';
import { GalleryScrollComponentPropsInterface } from '../component-props.interface';
import { GalleryScrollItemInterface } from '../item.interface';
import { TouchEventMock } from 'mocks/touch-event/mock';

const image1: GalleryScrollItemInterface = { sourceUrl: '1' };
const image2: GalleryScrollItemInterface = { sourceUrl: '2' };
const image3: GalleryScrollItemInterface = { sourceUrl: '3' };
const initialState3Images = galleryScrollMakeInitialState([image1, image2, image3], false);

describe('AppearOnScrollComponent', () => {
  let defaultProps: GalleryScrollComponentPropsInterface;
  let removeEventListenerMock: jest.Mock;

  beforeEach(() => {
    removeEventListenerMock = windowMockRemoveEventListener();

    defaultProps = {
      className: 'custom-class-name',
      items: [{ sourceUrl: '1' }, { sourceUrl: '2' }, { sourceUrl: '3' }, { sourceUrl: '4' }],
      isRtl: false,
      onTouch: jest.fn,
    };
  });

  it('should disconnect observer on unmount', async () => {
    reactMockUseReducer({ ...initialState3Images, pointerPositionStart: 20 });
    const { unmount } = render(<GalleryScrollComponent {...defaultProps} />);

    unmount();

    expect(removeEventListenerMock).toHaveBeenCalledTimes(5);
    expect(removeEventListenerMock.mock.calls[0][0]).toBe('contextmenu');
    expect(removeEventListenerMock.mock.calls[1][0]).toBe('mousemove');
    expect(removeEventListenerMock.mock.calls[2][0]).toBe('touchmove');
    expect(removeEventListenerMock.mock.calls[3][0]).toBe('mouseup');
    expect(removeEventListenerMock.mock.calls[4][0]).toBe('touchend');
  });

  it('should dispatch start on mouse down', () => {
    const { dispatchMock } = reactMockUseReducer({ ...initialState3Images, pointerPositionStart: 20 });

    const { getByTestId } = render(<GalleryScrollComponent {...defaultProps} />);

    fireEvent.mouseDown(getByTestId('GalleryScroll'), { clientX: 5 });

    expect(dispatchMock).toHaveBeenCalledTimes(1);
    expect(dispatchMock).toHaveBeenCalledWith({
      galleryLeft: 0,
      galleryRight: 0,
      positionX: 5,
      type: 'start',
    });
  });

  it('should dispatch start on touch start', () => {
    const { dispatchMock } = reactMockUseReducer({ ...initialState3Images, pointerPositionStart: 20 });

    const { getByTestId } = render(<GalleryScrollComponent {...defaultProps} />);

    fireEvent.touchStart(getByTestId('GalleryScroll'), {
      ...TouchEventMock(),
      changedTouches: [{ pageX: 5 } as Touch] as unknown as React.TouchList,
    });

    expect(dispatchMock).toHaveBeenCalledTimes(1);
    expect(dispatchMock).toHaveBeenCalledWith({
      galleryLeft: 0,
      galleryRight: 0,
      positionX: 5,
      type: 'start',
    });
  });

  it('single image case', () => {
    const { getAllByTestId } = render(<GalleryScrollComponent {...defaultProps} items={[image1]} />);

    expect(getAllByTestId('GalleryScrollPicture').length).toBe(1);
  });

  // TODO-FE[CX-408] Fix the gallery scroll image load test
  xit('loading class should be removed after image is loaded', () => {
    const { getByTestId } = render(<GalleryScrollComponent {...defaultProps} items={[image1]} />);

    fireEvent.load(getByTestId('GalleryScrollPicture'));

    expect(getByTestId('GalleryScrollPicture').className).toMatchInlineSnapshot(`"item"`);
  });
});
