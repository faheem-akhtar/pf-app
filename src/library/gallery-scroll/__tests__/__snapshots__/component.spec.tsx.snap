// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppearOnScrollComponent when gallery is touched, it should render all images 1`] = `"<div class=\\"custom-class-name\\"><div><picture style=\\"transition-duration: 0ms; transform: translateX(0%);\\" class=\\"loading\\"><source srcset=\\"1\\" type=\\"image/webp\\"><source srcset=\\"1\\" type=\\"image/jpeg\\"><img src=\\"1\\"></picture><picture style=\\"transition-duration: 0ms; transform: translateX(100%);\\" class=\\"loading\\"><source srcset=\\"2\\" type=\\"image/webp\\"><source srcset=\\"2\\" type=\\"image/jpeg\\"><img src=\\"2\\"></picture><picture style=\\"transform: translateX(-100%);\\" class=\\"loading\\"><source srcset=\\"3\\" type=\\"image/webp\\"><source srcset=\\"3\\" type=\\"image/jpeg\\"><img src=\\"3\\"></picture><picture style=\\"transition-duration: 0ms; transform: translateX(undefined%);\\" class=\\"loading\\"><source srcset=\\"4\\" type=\\"image/webp\\"><source srcset=\\"4\\" type=\\"image/jpeg\\"><img src=\\"4\\"></picture></div><div style=\\"width: 30px;\\"><div style=\\"transform: translateX(0px);\\"><div class=\\"\\"></div><div class=\\"\\"></div><div class=\\"\\"></div><div class=\\"undefined\\"></div></div></div></div>"`;
